### ISSUE TITLE
torch._foreach_add works failes to add constant to the list of tensors which leads to bugs in Adam

### ISSUE THREAD

**Issue by PgLoLo** [pytorch/pytorch#119885]
### 🐛 Describe the bug

I was debugging NaNs that occures in Adam  when some gradients contains zeros and I've managed to reduce the problem that `torch._foreach_add` somehow failes to add constant to all tensors in the list.
The following very small reproducer determenistically outputs wrong resutls:

```python
import torch as t
tensors = [t.zeros(shape).cuda() for shape in shapes]
tensors2 = torch._foreach_add(tensors, 1)
print(any((i.numel() > 0 and i.min() == 0) for i in tensors2))
```

So, we create list of zero tensors of some specific shapes (error does not accure on any shapes). Then we add `1` to them (not even 1e-8) and after then check if there are still zeros in the result. Unfortunately, this core returns True, which means "yes, _foreach_add failed to increment all tensors by one).

The exact shapes that reproduces this problem are the following (its quite large, that's why I put it under spoiler)

<details>

```python
shapes = [torch.Size([13, 256]),
 torch.Size([51, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([32, 256]),
 torch.Size([32]),
 torch.Size([256, 16]),
 torch.Size([256]),
 torch.Size([13, 256]),
 torch.Size([13, 256]),
 torch.Size([51, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 768]),
 torch.Size([768]),
 torch.Size([256, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256, 1024]),
 torch.Size([1024]),
 torch.Size([1024, 256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([256]),
 torch.Size([0, 256]),
 torch.Size([0])]
```
</details>

### Versions

I've managed to reproduce problem on all my machines. The most controllable environment in which this error can be reproduced is docker image `nvcr.io/nvidia/pytorch:23.05-py3`:

<details>

```
Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.35

Python version: 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-89-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 535.129.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             80
On-line CPU(s) list:                0-79
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
CPU family:                         6
Model:                              79
Thread(s) per core:                 2
Core(s) per socket:                 20
Socket(s):                          2
Stepping:                           1
CPU max MHz:                        3600.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           4390.13
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts md_clear flush_l1d
Virtualization:                     VT-x
L1d cache:                          1.3 MiB (40 instances)
L1i cache:                          1.3 MiB (40 instances)
L2 cache:                           10 MiB (40 instances)
L3 cache:                           100 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-19,40-59
NUMA node1 CPU(s):                  20-39,60-79
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT vulnerable

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] onnx==1.13.1rc2
[pip3] pytorch-quantization==2.1.2
[pip3] torch==2.0.0
[pip3] torch-tensorrt==1.4.0.dev0
[pip3] torchdata==0.6.0
[pip3] torchtext==0.15.1
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0
[conda] Could not collect
```
</details>

**Comment by janeyx99** [comment]
Ah, this is the same problem as https://github.com/pytorch/pytorch/issues/100701, and has since been resolved. This fix should have been in the latest release--could you update your torch version and try again?

**Comment by lancerts** [comment]
Tried on the recent release, the problem is fixed indeed..
```
import torch
shapes = [(2, 3), (3, 4), (4, 5)]
tensors = [torch.zeros(shape).cuda() for shape in shapes]
tensors2 = torch._foreach_add(tensors, 1)
print(tensors2)
print(any((i.numel() > 0 and i.min() == 0) for i in tensors2))
```
gives
```
(tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0'), tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]], device='cuda:0'), tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]], device='cuda:0'))
False

```

**Comment by PgLoLo** [comment]
> Ah, this is the same problem as #100701, and has since been resolved. This fix should have been in the latest release--could you update your torch version and try again?

You are absolutely correct! Currently problem is fixed. Huge relief, this problem was driving me nuts. Thank you!
